graylog2-server/src/main/java/org/graylog2/security/ldap/LdapConnector.java

```diff
@@ -99,32 +106,84 @@ public LdapEntry search(LdapNetworkConnection connection, String searchBase, Str
             LOG.trace("Search {} for {}, starting at {}",
                       activeDirectory ? "ActiveDirectory" : "LDAP", filter, searchBase);
         }
-        final EntryCursor entryCursor = connection.search(searchBase,
-                                                          filter,
-                                                          SearchScope.SUBTREE,
-                                                          "*");
-        final Iterator<Entry> it = entryCursor.iterator();
-        if (it.hasNext()) {
-            final Entry e = it.next();
-            // for generic LDAP use the dn of the entry for the subsequent bind, active directory needs the userPrincipalName attribute (set below)
-            if (!activeDirectory) {
+        EntryCursor entryCursor = null;
+        try {
+            entryCursor = connection.search(searchBase,
+                                            filter,
+                                            SearchScope.SUBTREE,
+                                            "*");
+            final Iterator<Entry> it = entryCursor.iterator();
+            if (it.hasNext()) {
+                final Entry e = it.next();
+                // always set the proper DN for the entry, we need it for group matching
                 ldapEntry.setDn(e.getDn().getName());
+
+                // for generic LDAP use the dn of the entry for the subsequent bind, active directory needs the userPrincipalName attribute (set below)
+                if (!activeDirectory) {
+                    ldapEntry.setBindPrincipal(e.getDn().getName());
+                }
+
+                for (Attribute attribute : e.getAttributes()) {
+                    if (activeDirectory && attribute.getId().equalsIgnoreCase("userPrincipalName")) {
```

Comment:
Inverse equality check (`"userPrincipalName".equalsIgnoreCase(…)`) to prevent (unlikely?) NPE.